Features/leaderboard

shaylasry opened the pull request.

lasrysha committed:
added all the needed scripts. still need to add the leaderboard gameobject

lasrysha committed:
working leaderboard with numbers ranks even if there is no score

lasrysha committed:
leaderboard visualization change

The rank number appears only if there is a score.
The first score that gets to the board stays - if the player get a score that is on the board its rank will be lower than the previous attempt/s with the same score

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/Leaderboard.cs:
```diff
@@ -0,0 +1,102 @@
+
+        for (position = 0; position < _leaderboardEntries.leaderboard.Count; position++)
+        {
+            if (_myScore > _leaderboardEntries.leaderboard[position].score) break;
```
you can just return position here, instead of the `break`

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Interfaces/IDataLoader.cs:
```diff
@@ -0,0 +1,4 @@
+﻿public interface IDataLoader
+{
+    public T LoadData<T>(string asset);
+}
```
let's combine this interface and the `IDataSaver` interface into a single interface called `DataHandler` which will contain two functions: `LoadData` and `SaveData`

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Interfaces/IDataSaver.cs:
```diff
@@ -0,0 +1,4 @@
+﻿public interface IDataSaver
+{
+    void SaveData<T>(string fileName, T data);
+}
```
see my comment on the `IDataLoader` file

IBLabs commented on Assets/BowAndArrowImproved/Scripts/JsonDataLoader.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using UnityEngine;
+
+public class JsonDataLoader : IDataLoader
+{
+    public T LoadData<T>(string asset)
+    {
+        try
+        {
+            T data = JsonUtility.FromJson<T>(asset);
+            return data;
+        }
+        catch (Exception e)
+        {
+            throw new Exception("Error loading Json data: " + e.Message);
+        }
+    }
+}
```
see my comment on `IDataLoader` interface

IBLabs commented on Assets/BowAndArrowImproved/Scripts/JsonDataSaver.cs:
```diff
@@ -0,0 +1,20 @@
+using System.IO;
+using UnityEngine;
+
+public class JsonDataSaver : IDataSaver
+{
+    public void SaveData<T>(string fileName, T data)
+    {
+        string jsonData = JsonUtility.ToJson(data);
+        string path = Path.Combine(Application.persistentDataPath, fileName);
+        try
+        {
+            File.WriteAllText(path, jsonData);
+        }
+        catch (Exception e)
+        {
+            throw new Exception("Error save to Json file: " + e.Message);
+        }
+    }
+}
```
see my comment on `IDataLoader` interface

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/Leaderboard.cs:
```diff
@@ -0,0 +1,102 @@
+    private const string JsonLeaderBoardFileName = "JsonLeaderboardHolder.json";
+    private LeaderboardEntries _leaderboardEntries;
+
+    private IDataLoader _leaderboardLoader = new JsonDataLoader();
+    private IDataSaver _leaderboardSaver = new JsonDataSaver();
```
that's a good case for a factory class, since `Leaderboard` doesn't really care what loader/saver it gets, it just want a data loader/saver, create class called `DataHandlerFactory` which contains a single `static` method called `get()`, the signature of this function should be something like:
```
public static IDataHandler get()
```
and inside it, create a `JsonDataHandler` instance and return it - then, here, simply use the factory class like so:
```
private IDataHander _dataHandler = DataHandlerFactory.get();
```

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/Leaderboard.cs:
```diff
@@ -0,0 +1,102 @@
+
+public class Leaderboard : MonoBehaviour
+{
+    private const string JsonLeaderBoardFileName = "JsonLeaderboardHolder.json";
```
change the name to snake case all caps so we know it's a constant and not a variable, like so: 
```
JSON_LEADERBOARD_FILE_NAME = ...
```

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/Leaderboard.cs:
```diff
@@ -0,0 +1,102 @@
+    {
+        if (newState == GameManager.State.Lose)
+        {
+            LoadLeaderboardData();
+
+            _myScore = scoreboard._score;
+            SubmitScore();
+
+            GenerateLeaderboardEntries();
+            gameObject.SetActive(true);
```
extract this to a different method called `HandleLoseState` or something like that, so it'll be easier to understand what's going on here and add more code to this event function in the future

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/Leaderboard.cs:
```diff
@@ -0,0 +1,102 @@
+
+        if (!File.Exists(filePath))
+        {
+            LeaderboardEntries leaderboardData = new LeaderboardEntries
+            {
+                leaderboard = new List<SingleLeaderboardEntry>()
+            };
+
+            string jsonDataToWrite = JsonUtility.ToJson(leaderboardData);
+
+            File.WriteAllText(filePath, jsonDataToWrite);
```
Extract this to a different function called `CreateLeaderboardDataFile()` or something like that

IBLabs commented on Assets/BowAndArrowImproved/Scripts/Leaderboard/ScoreEntry.cs:
```diff
@@ -0,0 +1,20 @@
+
+    public void InitEntryWithHighLight(string initRank, string initScore)
+    {
+        rank.text = $"<color=yellow>{initRank}</color>";
+        score.text = $"<color=yellow>{initScore}</color>";
```
wow cool, didn't know you can do that! 😲

lasrysha committed:
Reformat code

Merged IDataSaver and IDataLoader interfaces to IDataHandler, added DataHandlerFactory and brook down methods in leaderboard to smaller pieces

lasrysha committed:
added changes in leaderboard script and in json data holder to make the code more generic
